app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,437 @@
+    boolean gps_enabled = false;
+    boolean network_enabled = false;
+
+    try {
+      gps_enabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    try {
+      network_enabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    return (gps_enabled || network_enabled);
+  }
+
+  private void requestEnableLocationServices() {
+    alertDialogShower.show(KiwixDialog.EnableLocationServices.INSTANCE,
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            startActivityForResult(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
+                REQUEST_ENABLE_LOCATION_SERVICES);
+            return Unit.INSTANCE;
+          }
+        },
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            showToast(LocalFileTransferActivity.this, R.string.discovery_needs_location,
+                Toast.LENGTH_SHORT);
+            return Unit.INSTANCE;
+          }
+        });
+  }
+
+  private void requestEnableWifiP2pServices() {
+    alertDialogShower.show(KiwixDialog.EnableWifiP2pServices.INSTANCE,
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS));
+            return Unit.INSTANCE;
+          }
+        },
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            showToast(LocalFileTransferActivity.this, R.string.discovery_needs_wifi,
+                Toast.LENGTH_SHORT);
+            return Unit.INSTANCE;
+          }
+        });
+  }
+
+  @Override
+  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+    switch (requestCode) {
+      case REQUEST_ENABLE_LOCATION_SERVICES: {
+        LocationManager locationManager =
+            (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+
+        if (!(locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
+            || locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER))) {
+          // If neither provider is enabled
+          showToast(this, R.string.permission_refused_location, Toast.LENGTH_LONG);
+        }
+        break;
+      }
+
+      default: {
+        super.onActivityResult(requestCode, resultCode, data);
+        break;
+      }
+    }
+  }
+
+  /* Miscellaneous helper methods */
+  static void showToast(Context context, int stringResource, int duration) {
+    showToast(context, context.getString(stringResource), duration);
+  }
+
+  static void showToast(Context context, String text, int duration) {
+    Toast.makeText(context, text, duration).show();
+  }
+
+  @Override
+  public void onResume() {
+    super.onResume();
+    wifiDirectManager.registerWifiDirectBroadcastRecevier();
```

Comment:
I don't think the activity needs to know that wifiDirectManager has a broadcastReceiver. Also we don't want this in onResume/onPause, I probably confused the matter by talking about API symmetry.
If we unregister in onPause they we will not receive events that happen if the user has left our app but the transfer is still running, definitely not desired beahviour.

Reply:
> I don't think the activity needs to know that wifiDirectManager has a broadcastReceiver

So should I change the method name?

> Also we don't want this in onResume/onPause, I probably confused the matter by talking about API symmetry.
If we unregister in onPause they we will not receive events that happen if the user has left our app but the transfer is still running, definitely not desired beahviour.

I remember you had said this in a previous comment, but the thing is there's not a lot of state saving going on (or required)
Basically once the sender selects a receiver device, everything happens behind the scenes without any user intervention. Also, no wifi-p2p events are caused during the transfer by the app (unless the user turns off the wifi, which will cause an exception in the async-tasks and close LFTA).

Also, peer devices do not disappear immediately from the list returned by the API if unavailable (I found it took about 1-2 minutes for the device to be removed). So if the user navigates away for some time, registering the receiver afresh should help obtain a more accurate peer devices list from the framework)

Reply:
That is not how receivers work, they are active from the time they are registered to the time they are unregistered, they receive all events during that time so a "refresh" should have no effect because the act of registering should not cause any action[*].

>Also, no wifi-p2p events are caused during the transfer by the app (unless the user turns off the wifi, which will cause an exception in the async-tasks and close LFTA).

we can never make any guarantee about what will happen with a radio and a meddling user. Two years slaving on a Bluetooth medical app taught me that.

And on my first point it is still weird that a client class needs to tell WifiDirectManager about an implementation detail it should know nothing about, if you must call these methods then these yes you should change the name `startListentingToP2PEvents/stopListentingToP2PEvents` or something similar but I would reccomend these methods being deleted.

[*]Okay so there are some like when you register for battery updates, the intent returned by the Context.register call has the value of the battery in it, I think because it is a sticky broadcast but also cause it could just be "special". Mostly the returned intents are null or meaningless

Reply:
>  an implementation detail it should know nothing about

Didn't realise this before, but now it makes a lot of sense when I take into account that all the review changes have been towards making WDM's exposed parts as few as possible 
I have now [made it a part of](https://github.com/kiwix/kiwix-android/pull/1276/commits/32bfe70643db1b647479719fdc97595eaaa5ae21) `start/stopWifiDirectManager()` (which are called in onCreate & onDestroy)

> they receive all events during that time so a "refresh" should have no effect because the act of registering should not cause any action

Thanks, didn't know about this

> we can never make any guarantee about what will happen with a radio and a meddling user

I have a query though: The code currently doesn't listen for the `WIFI_P2P_DISCOVERY_CHANGED_ACTION` broadcast. Should I add a log in the `default` case of the switch for this (& other unandled) broadcasts? For debugging purposes

Reply:
I am generally against code existing in production that is "useful for debugging" when you could just attach a debugger and add a breakpoint or a conditional breakpoint.
Overall this doesn't hurt the quality overtly so if you think it useful then do so.

Reply:
I'll leave it be - there's no need for the log, it will only clutter the code
I think this review can be resolved now, since
> I have now made it a part of start/stopWifiDirectManager() (which are called in onCreate & onDestroy)

So LFTA doesn't know about the broadcast receiver, and it exists from onCreate to onDestroy